examples/c++/userPropagator.cpp

```diff
@@ -0,0 +1,420 @@
+#include <algorithm>
+#include <chrono>
+#include <iostream>
+#include <random>
+#include <stack>
+#include <unordered_map>
+#include <unordered_set>
+#include <vector>
+#include <cstring>
+#include "z3++.h"
+
+/**
+ * The program solves the n-queens problem (number of solutions) with 4 different approaches
+ * 1) Bit-Vector constraints + Default solver + Blocking Clauses
+ * 2) Bit-Vector constraints + Simple solver + Blocking Clauses
+ * 3) Bit-Vector constraints + Simple solver + Adding contradictions in the propagator
+ * 4) Constraints only implicit via the propagator + Simple solver + Adding contradictions in the propagator
+ *
+ * Runs 1 + 2 are done for comparision with 3 + 4
+ */
+
+using namespace std::chrono;
+using std::to_string;
+
+#define QUEEN
+#define REPETITIONS 5
+
+#define SIZE(x) std::extent<decltype(x)>::value
+
+#if LOG
+#define WriteEmptyLine std::cout << std::endl
+#define WriteLine(x) std::cout << (x) << std::endl
+#define Write(x) std::cout << x
+#else
+#define WriteEmptyLine
+#define WriteLine(x)
+#define Write(x)
+#endif
+
+class model {
```

Comment:
I don't get why you need a model abstraction.
Why not class model : public std::vector<unsigned> {};